### PR TITLE
add code comment style guide for TiKV

### PR DESCRIPTION
**Pull request by dcalvin**
Signed-off-by: Calvin Weng <wenghao@pingcap.com>

## What have you changed?  

- Add code comment style guide for TiKV repo

## What are the type of the changes?  

- Improvement

### CODE REVIEW

File: code-comment-style.md [hunk 1]
```diff
@@ -0,0 +1,137 @@
...
+    - Use **American English** rather than British English
+        
+        - color, canceling, synchronize     (Recommended)
+        - colour, cancelling, synchronise   (Not recommended)
+    
+    - Use correct spelling
+
+    - Use **standard or official capitalization**
+        
+        - TiKV, TiDB-Binlog, Region, gRPC, RocksDB, GC, k8s, [mydumper](https://github.com/maxbube/mydumper), [Prometheus Pushgateway](https://github.com/prometheus/pushgateway)   (Right)
```
Comment (siddontang) [inline code-comment-style.md]: K8s
not k8s

Comment (siddontang) [reply code-comment-style.md]: https://kubernetes.io/docs/concepts/overview/what-is-kubernetes/#what-does-kubernetes-mean-k8s

File: code-comment-style.md [hunk 2]
```diff
@@ -0,0 +1,137 @@
...
+
+    - Use standard grammar and correct punctuation
+    - Use relatively short sentences
+
+- For each comment, capitalize the first letter and end this sentence with a period
+    
+    - If a lower-case identifier comes at the beginning of a sentence, don't capitalize it
+
+    ```
+    // enterGame causes Players to enter the
```
Comment (siddontang) [inline code-comment-style.md]: For Rust, it is not recommended to use the function name as the start word in comments.

**Comment by siddontang** [comment]
Please give your advice. @nrc

File: code-comment-style.md (hunk 1 above)
Comment (dcalvin) [reply code-comment-style.md]: Good spot. Let's stick with K8s in our case.

File: code-comment-style.md (hunk 2 above)
Comment (dcalvin) [reply code-comment-style.md]: Got it. I can add a note to it.  Here it does not specifically talk about function name. An identifier is a generic word that refers to any construction element like a variable, function, or method. In rust, can we use other names as the start words besides function?

File: code-comment-style.md (hunk 1 above)
Comment (CaitinChen) [reply code-comment-style.md]: @siddontang @dcalvin “Kubernetes” was stylized as “K8s” formerly; but now it is “k8s”.

Comment (dcalvin) [reply code-comment-style.md]: As later confirmed with CC, k8s in all lower-cased form is generally accepted. Only initial-capped as K8s when it is the first word of the sentence.

Comment (dcalvin) [reply code-comment-style.md]: As later confirmed with CC, k8s in all lower-cased form is generally accepted. Only initial-capped as K8s when it is the first word of the sentence.

File: code-comment-style.md [hunk 3]
```diff
@@ -0,0 +1,137 @@
...
+- Exported name
+- Test case
+- TODO
+- FIXME
+
+## How to comment?
+
+### Format of a good comment (for Rust)
+
+- Non-doc comment
```
Comment (Connor1996) [inline code-comment-style.md]: When to use non-doc comment and when not to use?

Comment (dcalvin) [reply code-comment-style.md]: As a general thumb of rule, use doc comment as long as you want to refer to that information in the command line. In our case, we have the convention to use doc comment for (public or private) structures including the fields, macros, methods, etc.  We use non-doc (code) comment for everything else like inner steps of functions.

Will add this in the comment style guide.

File: code-comment-style.md (hunk 2 above)
Comment (dcalvin) [reply code-comment-style.md]: Will remove this since as confirmed by CC, this (identifier) is actually more intended for  Go.

File: code-comment-style.md [hunk 4]
```diff
@@ -0,0 +1,128 @@
...
+    - Used for inline code comment (for better code explaination when reviewing)
+    - Use **//** for a line comment
+    - Use **/* … */** for a block comment (used only when needed)
+    
+- Doc comment
+
+    - Used in module level descriptions (structures, fields, macros, etc.) for command line reference
+    - Use **///** for a line comment 
+    - Use **//!** for a block comment
+    - Use **rustfmt** to format your code
```
Comment (nrc) [inline code-comment-style.md]: I would not use rustfmt to format comments - it is not stable for comments and formats imperfectly

File: code-comment-style.md [hunk 5]
```diff
@@ -0,0 +1,128 @@
...
+
+## How to comment?
+
+### Format of a good comment (for Rust)
+
+- Non-doc comment
+
+    - Used for inline code comment (for better code explaination when reviewing)
+    - Use **//** for a line comment
+    - Use **/* … */** for a block comment (used only when needed)
```
Comment (nrc) [inline code-comment-style.md]: I would never use block comments

File: code-comment-style.md (hunk 3 above)
Comment (nrc) [reply code-comment-style.md]: I would state the distinction as "use doc comments to document the interface of code and use non-doc comments to document the implementation details"

File: code-comment-style.md [hunk 6]
```diff
@@ -0,0 +1,128 @@
...
+
+- File
+- Type
+- Constant
+- Function
+- Method
+- Variable
+- Typical algorithm
+- Exported name
+- Test case
```
Comment (nrc) [inline code-comment-style.md]: My preference is not to require comments anywhere since that leads to boilerplate. Use comments when there is context that is missing from the code or is hard to deduce from the code. For example, what are the side-effects of calling a function, or the performance characteristics, or the use cases which the code addresses.

As an explicit counter-example I think comments like:

```
/// Returns a record from the database.
pub fn get_record(db: &Database) -> Record { ... }
```

are a negative and should be avoided, better to have no comment at all.

A good comment is something like:

```
/// This function may get the record from disk which means it will block waiting for IO. It is nearly always
/// better to call `async_get_record` unless you are sure that the requested record is cached in memory.
///
/// If you need to get many records, calling this function many times has O(n) performance. Consider using
/// `get_records` which has O(log n) performance.
///
/// If `db` is in an inconsistent state, this function will panic.
pub fn get_record(db: &Database) -> Record { ... }
```

Note that we don't say what `get_record` does but describe aspects of how it should be used which are not obvious from the function signature.

File: code-comment-style.md [hunk 7]
```diff
@@ -0,0 +1,128 @@
...
+- For obscure code
+- For tricky or interesting code
+- For a complex code block
+- If a bug exists in the code but you cannot fix it or you just want to ignore it for the moment
+- If the code is not optimal but you don’t have a smarter way now
+- To remind yourself or others of missing functionality or upcoming requirements not present in the code
+
+A comment is generally required for:
+
+- File
```
Comment (nrc) [inline code-comment-style.md]: Prefer module to file, since files are a secondary concept in Rust

File: code-comment-style.md [hunk 8]
```diff
@@ -0,0 +1,128 @@
...
+- Doc comment
+
+    - Used in module level descriptions (structures, fields, macros, etc.) for command line reference
+    - Use **///** for a line comment 
+    - Use **//!** for a block comment
+    - Use **rustfmt** to format your code
+
+- Place the single-line and block comment above the code it’s annotating
+- Fold long lines of comments
+- Each line of text in your code and comment should be at most 100 characters long
```
Comment (nrc) [inline code-comment-style.md]: I prefer 80, 100 is fine for code, but I think is a long line of text to read (studies show fewer words per column are easier to read).

File: code-comment-style.md [hunk 9]
```diff
@@ -0,0 +1,128 @@
...
+        - TiKV, TiDB-Binlog, Region, gRPC, RocksDB, GC, k8s, [mydumper](https://github.com/maxbube/mydumper), [Prometheus Pushgateway](https://github.com/prometheus/pushgateway)   (Right)
+        - Tikv, TiDB Binlog, region, grpc, rocksdb, gc, k8S, MyDumper, Prometheus PushGateway   (Wrong)
+
+    - Use words and expressions consistently
+        
+        - dead link vs. broken link (Only one of them can appear in a single document)
+    
+    - Do not use lengthy compound words
+
+    - Do not abbreviate unless it is absolutely necessary
```
Comment (nrc) [inline code-comment-style.md]: I think abreviations can be fine if they improve readability, but avoid overusing them. For example, it would be tedious to read 'Titanium Data Base' all the time instead of 'TiDB', or to spell out 'HTTP'

File: code-comment-style.md [hunk 10]
```diff
@@ -0,0 +1,128 @@
...
+
+    - "we" can be used only when it means the code writer
```
Comment (nrc) [inline code-comment-style.md]: Should the author always prefer 'we' to 'I'? (I think yes)

File: code-comment-style.md [hunk 11]
```diff
@@ -0,0 +1,128 @@
...
+    
+    - Opens the `log` file  
+
+### Tips for a good comment
+
+- Comment code while writing it
+- Do not assume the code is self-evident
+- Avoid unnecessary comments for simple code
+- Write comments as if they were for you
+- Make sure the comment is up-to-date
```
Comment (nrc) [inline code-comment-style.md]: I would also add "Make sure you keep comments up to date when you edit code"

File: code-comment-style.md (hunk 2 above)
Comment (nrc) [reply code-comment-style.md]: I would add that I would always use markdown code for names of functions, etc., so that if you *have* to start a sentence with a name, then it looks more reasonable, e.g.,

```
// `enter_game` causes Player to enter the ...
fn enter_game ...
```

File: code-comment-style.md (hunk 10 above)
Comment (dcalvin) [reply code-comment-style.md]: Agree. It would be self-pretentious to use "I".

File: code-comment-style.md (hunk 9 above)
Comment (dcalvin) [reply code-comment-style.md]: True. Since this is code comment we do not have to follow this rule as strictly as documentation. For general accepted/known abbreviations, we can and should use to improve readability. For less-known items, we may still need to spell out

File: code-comment-style.md (hunk 6 above)
Comment (dcalvin) [reply code-comment-style.md]: Good point. Will add in style guide. We should not comment for comment purposes, but comment to deliver something that matters

File: code-comment-style.md (hunk 4 above)
Comment (dcalvin) [reply code-comment-style.md]: Noted.

Comment (dcalvin) [reply code-comment-style.md]: @zhangjinpeng1987 PTAL

File: code-comment-style.md (hunk 8 above)
Comment (BusyJay) [reply code-comment-style.md]: I suggest to keep the same value as [style guide](https://github.com/rust-dev-tools/fmt-rfcs/blob/master/guide/guide.md#indentation-and-line-width). And I prefer 100, which works better for my large monitor. :rofl:

File: code-comment-style.md (hunk 4 above)
Comment (dcalvin) [reply code-comment-style.md]: Here **rustfmt** is used to format code, not the comment, so will leave as is for now.

File: code-comment-style.md (hunk 5 above)
Comment (dcalvin) [reply code-comment-style.md]: Not contradicted with what's implied by "used only when needed"

File: code-comment-style.md [hunk 12]
```diff
@@ -0,0 +1,129 @@
+# Code Comment Style
+
+This document describes the code comment style applied to TiKV repositories. When you are to commit, be sure to follow the style to write good code comments.
+
+## Why a good comment matters?
```
Comment (Hoverbear) [inline code-comment-style.md]:
```suggestion
## Why does a good comment matter?
```

File: code-comment-style.md [hunk 13]
```diff
@@ -0,0 +1,129 @@
...
+
+    - Used to document implementation details
+    - Use **//** for a line comment
+    - Use **/* … */** for a block comment (used only when needed)
+    
+- Doc comment
+
+    - Used to document the interface of code (structures, fields, macros, etc.)
+    - Use **///** for a line comment 
+    - Use **//!** for a block comment
```
Comment (Hoverbear) [inline code-comment-style.md]:
```suggestion
    - Use **//!** for module level documentation```

File: code-comment-style.md [hunk 14]
```diff
@@ -0,0 +1,129 @@
...
+- Non-doc comment
+
+    - Used to document implementation details
+    - Use **//** for a line comment
+    - Use **/* … */** for a block comment (used only when needed)
+    
+- Doc comment
+
+    - Used to document the interface of code (structures, fields, macros, etc.)
+    - Use **///** for a line comment
```
Comment (Hoverbear) [inline code-comment-style.md]:
```suggestion
    - Use **///** for item documentation (functions, attributes, structures, etc)
```

File: code-comment-style.md (hunk 13 above)
Comment (Hoverbear) [reply code-comment-style.md]: (There is also block comments but according to @brson they are bad style. They are `/*! */`.)

File: code-comment-style.md [hunk 15]
```diff
@@ -0,0 +1,129 @@
...
+    - Used to document implementation details
+    - Use **//** for a line comment
+    - Use **/* … */** for a block comment (used only when needed)
+    
+- Doc comment
+
+    - Used to document the interface of code (structures, fields, macros, etc.)
+    - Use **///** for a line comment 
+    - Use **//!** for a block comment
+    - Use **rustfmt** to format your code
```
Comment (Hoverbear) [inline code-comment-style.md]: Not sure this is appropriate for a comment style document. Rustfmt is only for code.

File: code-comment-style.md [hunk 16]
```diff
@@ -0,0 +1,129 @@
...
+
+- Place the single-line and block comment above the code it’s annotating
+- Fold long lines of comments
+- The maximum width for a line is 100 characters
+- Do not use relative URLs; use **absolute URLs** instead
```
Comment (Hoverbear) [inline code-comment-style.md]: It is absolutely appropriate to use relative URLs in autogenerated documentation. The user's machine will generate them via `cargo doc` and the hosting location may be anything. So using an absolute URL will either break local documentation, or hosted documentation.

For example, this is how you link to another struct: https://github.com/tikv/client-rust/blob/master/src/raw.rs#L75

File: code-comment-style.md [hunk 17]
```diff
@@ -0,0 +1,129 @@
...
+
+    - Use words and expressions consistently
+        
+        - dead link vs. broken link (Only one of them can appear in a single document)
+    
+    - Do not use lengthy compound words
+
+    - Do not abbreviate unless it is necessary (for readability purposes)
+
+    - "we" can be used only when it means the code writer
```
Comment (Hoverbear) [inline code-comment-style.md]:
```suggestion
    - "we" should be used only when it means the code writer *and* the reader.
```

File: code-comment-style.md (hunk 8 above)
Comment (dcalvin) [reply code-comment-style.md]: Updated as suggested by Jay.

File: code-comment-style.md (hunk 15 above)
Comment (dcalvin) [reply code-comment-style.md]: Decided to remove this after discussion with Catin, since it does not pertain to comment style guide.

File: code-comment-style.md (hunk 16 above)
Comment (dcalvin) [reply code-comment-style.md]: As confirmed with @BusyJay, we should use relative URLs for local links.  Now updated as: Use "Use relative URLs when you link to positions on your local machine"

**Comment by dcalvin** [comment]
> Looks nice. I think we should also link to this file in the contribution documentation?

Sure thing. Will update when this one is approved.

**Comment by breeswish** [comment]
May be you can just update them together in this PR because adding a single line referring to the doc is not a matter.

File: code-comment-style.md [hunk 18]
```diff
@@ -0,0 +1,129 @@
+# Code Comment Style
```
Comment (breeswish) [inline code-comment-style.md]: Maybe this file need to be capitalized like other md files in the project root?

Comment (dcalvin) [reply code-comment-style.md]: Good point. Will update accordingly.

**Comment by brson** [comment]
I'll read this tomorrow.

File: CODE_COMMENT_STYLE.md [hunk 19]
```diff
@@ -0,0 +1,116 @@
...
+- Doc comment
+
+    - Used to document the interface of code (structures, fields, macros, etc.)
+    - Use **///** for item documentation (functions, attributes, structures, etc)
+    - Use **//!** for module level documentation
+
+- Place the single-line and block comment above the code it’s annotating
+- Fold long lines of comments
+- The maximum width for a line is 100 characters
+- Use relative URLs when you link to positions on you local machine
```
Comment (Hoverbear) [inline CODE_COMMENT_STYLE.md]:
```suggestion
- Use relative URLs when appropriate for Rustdoc links
```

File: CODE_COMMENT_STYLE.md [hunk 20]
```diff
@@ -0,0 +1,116 @@
...
+- Type
+- Constant
+- Function
+- Method
+- Variable
+- Typical algorithm
+- Exported name
+- Test case
+- TODO
+- FIXME
```
Comment (brson) [inline CODE_COMMENT_STYLE.md]: Some of thee seem too demanding to me, and I might separate this into "comments" and "doc comments"; with doc comments required for: _public_ modules, types, constants, functions, methods, fields (which could be summarized a "all public items (such as modules, types, constants, functions, and fields) that are exported from the crate". That's a pretty typical Rust policy.

Then the remaining are sort of PingCAP policies, some of which I'd take issue with. For example, I would not put a comment on most variables, preferring the variable name be descriptive enough to understand its purpose from the code. It's also not clear to me what "typical algorithm" means, though I might interpret this as just "algorithms", or maybe "Identify the name of known algorithms", and "describe complex algorithms".

I'm not sure about the "exported name" case. Is this a rustdoc thing, or a tikv-thing? (I mostly don't comment `pub use`. Though maybe this just refers to any "thing" that's exported? It could be clearer.

Please give a thought to the above and see if things can be clarified a bit. I'm happy to offer suggestions if needed.

File: CODE_COMMENT_STYLE.md [hunk 21]
```diff
@@ -0,0 +1,116 @@
...
+
+## How to comment?
+
+### Format of a good comment (for Rust)
+
+- Non-doc comment
+
+    - Used to document implementation details
+    - Use **//** for a line comment
+    - Use **/* … */** for a block comment (used only when needed)
```
Comment (brson) [inline CODE_COMMENT_STYLE.md]: I don't see why the rules about block comments should be different for doc comments and non-doc comments. Both cases are aesthetic choices for the eyes of maintainers.

I'm sympathetic to block comments, but I think it would be more consistent to generally disallow block comments everywhere. Maybe with a note indicated possible exception cases, like for commenting out code that for some reason still needs to be checked in; or for cases where you _really_ _really_ want to write with block comments for personal reasons; or temporarily for drafting long-form comments, prior to converting them to line comments.

File: CODE_COMMENT_STYLE.md [hunk 22]
```diff
@@ -0,0 +1,116 @@
...
+    - Use correct spelling
+
+    - Use **standard or official capitalization**
+        
+        - TiKV, TiDB-Binlog, Region, gRPC, RocksDB, GC, k8s, [mydumper](https://github.com/maxbube/mydumper), [Prometheus Pushgateway](https://github.com/prometheus/pushgateway)   (Right)
+        - Tikv, TiDB Binlog, region, grpc, rocksdb, gc, k8S, MyDumper, Prometheus PushGateway   (Wrong)
+
+    - Use words and expressions consistently
+        
+        - dead link vs. broken link (Only one of them can appear in a single document)
```
Comment (brson) [inline CODE_COMMENT_STYLE.md]: Use quotes here "dead link" vs. "broken link"

File: CODE_COMMENT_STYLE.md [hunk 23]
```diff
@@ -0,0 +1,116 @@
...
+    - Do not abbreviate unless it is necessary (for readability purposes)
+
+    - "we" should be used only when it means the code writer *and* the reader
+
+- Sentence
+
+    - Use standard grammar and correct punctuation
+    - Use relatively short sentences
+
+- For each comment, capitalize the first letter and end this sentence with a period
```
Comment (brson) [inline CODE_COMMENT_STYLE.md]: "this sentence" -> "the sentence"

Though I might reword this as "Capitalize the first letter of sentences and end them with periods".

Comment (brson) [reply CODE_COMMENT_STYLE.md]: I note though that _all_ of the bullet points in this document (which is mostly bullet points) do not end with any punctuation, which to the reader may _appear_ contradictory to this recommendation.

File: CODE_COMMENT_STYLE.md (hunk 20 above)
Comment (dcalvin) [reply CODE_COMMENT_STYLE.md]: Great inputs. 
For typical rust items, will add a general description as suggested
For common items that look confusing:

- Will not use "required" since it's denoting mandatory instead of necessary.  So no pressure.
-  Removed "exported name" since its a golang thing.
- "complex algorithm" instead of "typical

File: CODE_COMMENT_STYLE.md (hunk 21 above)
Comment (dcalvin) [reply CODE_COMMENT_STYLE.md]: Updated as:   
 **Note**: Block comments are not recommended unless for personal reasons or temporary purposes prior to being converted to line comments)

File: CODE_COMMENT_STYLE.md [hunk 24]
```diff
@@ -0,0 +1,116 @@
...
+- FIXME
+
+## How to comment?
+
+### Format of a good comment (for Rust)
+
+- Non-doc comment
+
+    - Used to document implementation details
+    - Use **//** for a line comment
```
[the remark this answers is not in the document]
Comment (Hoverbear) [reply CODE_COMMENT_STYLE.md]: `//`, even.

File: CODE_COMMENT_STYLE.md (hunk 21 above)
[the remark this answers is not in the document]
Comment (Hoverbear) [reply CODE_COMMENT_STYLE.md]: `/* ... */`, even.

File: CODE_COMMENT_STYLE.md [hunk 25]
```diff
@@ -0,0 +1,116 @@
...
+    - Use **/* … */** for a block comment (used only when needed)
+    
+- Doc comment
+
+    - Used to document the interface of code (structures, fields, macros, etc.)
+    - Use **///** for item documentation (functions, attributes, structures, etc)
```
[the remark this answers is not in the document]
Comment (Hoverbear) [reply CODE_COMMENT_STYLE.md]: `///`, even.

File: CODE_COMMENT_STYLE.md [hunk 26]
```diff
@@ -0,0 +1,116 @@
...
+    - Use **//!** for module level documentation
```
[the remark this answers is not in the document]
Comment (Hoverbear) [reply CODE_COMMENT_STYLE.md]: `//!`, even.

File: CODE_COMMENT_STYLE.md (hunk 24 above)
[the remark this answers is not in the document]
Comment (dcalvin) [reply CODE_COMMENT_STYLE.md]: "even" meaning?

[the remark this answers is not in the document]
Comment (Hoverbear) [reply CODE_COMMENT_STYLE.md]: "Alternatively, `//`"

**Comment by Hoverbear** [comment]
@dcalvin This commit: https://github.com/tikv/tikv/pull/4204/commits/7e0f60125e44c0093496f9d77a9903f72d2067b3 isn't signoff'd. So we can't merge. :(

**Comment by dcalvin** [comment]
@siddontang Is there a way to manually set DCO passing for this PR? Obviously, there are one or two commits somehow missing DCO that block the merging.

**Comment by brson** [comment]
Superceded by https://github.com/tikv/tikv/pull/4259
